## Datalab Semester 2

In [1]:
import os
import sqlite3
import pandas as pd

# 1. Zoek de map op waar dit notebook-bestand staat
map_van_notebook = os.path.dirname(os.path.abspath('__file__'))

# 2. Maak het volledige pad naar de database
db_pad = os.path.join(map_van_notebook, 'database.sqlite')
conn = sqlite3.connect(db_pad)

In [2]:
tabellen = pd.read_sql_query("SELECT name FROM sqlite_master WHERE type='table'", conn)
tabellen


,name
0,sqlite_sequence
1,Player_Attributes
2,Player
3,Match
4,League
5,Country
6,Team
7,Team_Attributes


In [3]:
query_1 = "SELECT * FROM sqlite_sequence"

df_sqlite_sequence = pd.read_sql_query(query_1, conn)
df_sqlite_sequence.head()

query_2 = "SELECT * FROM player_Attributes"

df_player_Attribute = pd.read_sql_query(query_2, conn)
df_player_Attribute.head()

query_3 = "SELECT * FROM Player"

df_player = pd.read_sql_query(query_3, conn)
df_player.head()

query_4 = "SELECT * FROM Match"

df_match = pd.read_sql_query(query_4, conn)
df_match.head()

query_5 = "SELECT * FROM League"

df_league = pd.read_sql_query(query_5, conn)
df_league.head()

query_6 = "SELECT * FROM Country"

df_country = pd.read_sql_query(query_6, conn)
df_country.head()

query_7 = "SELECT * FROM Team_Attributes"

df_team_attributes = pd.read_sql_query(query_7, conn)
df_team_attributes.head()

,id,team_fifa_api_id,team_api_id,date,buildUpPlaySpeed,buildUpPlaySpeedClass,buildUpPlayDribbling,buildUpPlayDribblingClass,buildUpPlayPassing,buildUpPlayPassingClass,...,chanceCreationShooting,chanceCreationShootingClass,chanceCreationPositioningClass,defencePressure,defencePressureClass,defenceAggression,defenceAggressionClass,defenceTeamWidth,defenceTeamWidthClass,defenceDefenderLineClass
0,1,434,9930,2010-02-22 00:00:00,60,Balanced,NaN,Little,50,Mixed,...,55,Normal,Organised,50,Medium,55,Press,45,Normal,Cover
1,2,434,9930,2014-09-19 00:00:00,52,Balanced,48.0,Normal,56,Mixed,...,64,Normal,Organised,47,Medium,44,Press,54,Normal,Cover
2,3,434,9930,2015-09-10 00:00:00,47,Balanced,41.0,Normal,54,Mixed,...,64,Normal,Organised,47,Medium,44,Press,54,Normal,Cover
3,4,77,8485,2010-02-22 00:00:00,70,Fast,NaN,Little,70,Long,...,70,Lots,Organised,60,Medium,70,Double,70,Wide,Cover
4,5,77,8485,2011-02-22 00:00:00,47,Balanced,NaN,Little,52,Mixed,...,52,Normal,Organised,47,Medium,47,Press,52,Normal,Cover


Beschrijving van de Dataset

Wij hebben één dataset met acht tabellen gekregen en sommige kolommen van onze datasets hebben een relatie met elkaar. Namelijk id; dit is een unieke waarde binnen een tabel en deze hebben we in elke tabel. De volgende kolommen hebben een relatie met de tabellen Teams en Team_Attributes: team_api_id bevat unieke identifiers van een team en die is van groot belang voor ons. team_long_name bevat de volledige namen van teams. team_short_name bevat een afkorting van de volledige naam van een team en kan soms ook belangrijk zijn. Dat waren ook alle kolommen uit de tabel Team.

De tabel Team_Attributes heeft nog een belangrijke kolom die niet met de Team-tabel overlapt, namelijk date; deze geeft de momenten weer waarop een speelstijl gemeten is. Alle andere kolommen bevatten speelstijl-variabelen. Opbouw of build-up: bij opbouw horen buildUpPlaySpeed, buildUpPlayPassing en buildUpPlayDribbling. Aanval of chance creation: bij aanval horen de kolommen chanceCreationPassing, chanceCreationCrossing en chanceCreationShooting. Verdediging of defence: bij verdediging horen de kolommen defencePressure, defenceAggression en defenceTeamWidth.

Overlappingen komen ook voor tussen de tabellen Player en Player_Attributes. player_api_id is een unieke identifier van een speler en player_fifa_api_id is een unieke FIFA-identifier. De tabel Player bevat daarnaast de kolom player_name, die de naam van een speler geeft; dit is zeer belangrijk voor ons. birthday geeft een specifieke geboortedatum aan; dit is van belang omdat leeftijd voor ons een van de cruciale factoren is. height geeft de lengte van een speler aan; dit kan belangrijk zijn voor het controleren van de gezondheid, en om dezelfde reden is de kolom weight voor ons relevant.

De tabel Player_Attributes bevat ook kolommen. Date geeft het moment van de meting weer; dit kan belangrijk voor ons zijn. De volgende kolommen kunnen allemaal van belang zijn en we kunnen ze in groepen verdelen. Bij Performance rating horen de kolommen overall_rating en potential. Bij Speelstijl horen preferred_foot, attacking_work_rate en defensive_work_rate. Bij Skills horen crossing, vision, penalties, marking, standing_tackle en sliding_tackle. Bij Keeper stats horen gk_diving, gk_handling en gk_reflexes; deze kolommen zijn alleen van toepassing op keepers en analyse daarvoor 

Bij de tabellen Match en League overlapt de kolom country_id; dit is een unieke ID van een land en deze is erg belangrijk. De tabel League heeft nog een kolom, namelijk name, die de naam van een league geeft. 

In de tabel Match staat ook de kolom league_id, die een unieke ID van een league bevat, deze hebben dus dan relatie metelkaar. Season geeft de verschillende jaartallen van de seizoenen weer; dit kan ook belangrijk zijn voor het filteren op bepaalde seizoen. match_api_id geeft de unieke ID van een wedstrijd. Stage geeft de spelronde aan; dit is belangrijk voor ons om de conditie van een speler te meten. We kunnen hier weer verschillende kolommen in groepen verdelen. Bij Teams in de wedstrijd horen home_team_api_id en away_team_api_id. Bij Resultaat horen home_team_goal en away_team_goal. Bij Odds horen de variabelen SJA, VCH, VCD, VCA, GBH, GBD, GBA, BSH, BSD en BSA.

De tabel Country heeft nog een kolom name, die de naam van een land doorgeef. En als laatste: de tabel sqlite_sequence is de enige die geen id heeft, maar wel een kolom name, die bestaat uit de namen van alle tabellen die we hier gebruiken, en seq geeft het aantal sequences aan. Beide zijn belangrijk om de data beter in kaart te houden.

In [4]:
query = "SELECT * FROM Team"

df_teams = pd.read_sql_query(query, conn)
df_teams.head()

,id,team_api_id,team_fifa_api_id,team_long_name,team_short_name
0,1,9987,673.0,KRC Genk,GEN
1,2,9993,675.0,Beerschot AC,BAC
2,3,10000,15005.0,SV Zulte-Waregem,ZUL
3,4,9994,2007.0,Sporting Lokeren,LOK
4,5,9984,1750.0,KSV Cercle Brugge,CEB


In [5]:
# Automatisch het juiste pad vinden, ongeacht waar je bent op je Mac
current_dir = os.path.dirname(os.path.abspath('__file__'))
db_path = os.path.join(current_dir, 'database.sqlite')

print(f"Ik zoek de database op: {db_path}")

conn = sqlite3.connect(db_path)
query = "SELECT * FROM Team WHERE team_long_name LIKE '%Barcelona%'"

df = pd.read_sql_query(query, conn)
df.head()


Ik zoek de database op: /Users/emirsmacbookair/Documents/GitHub/Datalab_semester2_Groep1/notebook/database.sqlite


,id,team_api_id,team_fifa_api_id,team_long_name,team_short_name
0,43042,8634,241,FC Barcelona,BAR


Team_api_id is het unieken id nummer van elk team en heeft relaties binnen andere tabellen in de database, bijvoorbeeld nummer 8634 is unieke id van team FC barcelona, dit zorgt voor duidelijkheid binnen het database en zorgt alles makkelijk te lezen. nog een voorbeeld is dat als we binnen tabel matches alleen van team fc barcelona willen ophalen dan kan filteren op de unieke team_api_id. dit maakt analyses makkelijker en zichtbaar.

Ranglijst opstellen

In [6]:

def bepaal_match_punten(row):
    """
    Berekent de punten voor de thuis- en uitploeg op basis van de wedstrijdscore.
    
    Args:
        row (pd.Series): Een rij uit de Match dataframe met 'home_team_goal' en 'away_team_goal'.
        
    Returns:
        pd.Series: De behaalde punten voor [home_points, away_points].
    """
    if row['home_team_goal'] > row['away_team_goal']:
        return pd.Series([3, 0], index=['home_points', 'away_points'])
    elif row['home_team_goal'] < row['away_team_goal']:
        return pd.Series([0, 3], index=['home_points', 'away_points'])
    else:
        return pd.Series([1, 1], index=['home_points', 'away_points'])

def genereer_competitie_ranglijst(connection, seizoen):
    """
    Haalt wedstrijddata op uit de database en genereert een gesorteerde ranglijst.
    
    Args:
        connection (sqlite3.Connection): De actieve database verbinding.
        seizoen (str): Het gewenste seizoen (bijv. '2015/2016').
        
    Returns:
        pd.DataFrame: Een dataframe met de teamnamen en hun totale punten, gesorteerd van hoog naar laag.
    """
    # 1. Data ophalen
    query = f"SELECT home_team_api_id, away_team_api_id, home_team_goal, away_team_goal FROM Match WHERE season = '{seizoen}'"
    df_matches = pd.read_sql_query(query, connection)
    
    # 2. Punten berekenen met de hulpfunctie
    df_matches[['home_points', 'away_points']] = df_matches.apply(bepaal_match_punten, axis=1)
    
    # 3. Groeperen en totalen berekenen
    home_stats = df_matches.groupby('home_team_api_id')['home_points'].sum().reset_index()
    away_stats = df_matches.groupby('away_team_api_id')['away_points'].sum().reset_index()
    
    home_stats.columns = ['team_api_id', 'points']
    away_stats.columns = ['team_api_id', 'points']
    
    ranglijst = pd.concat([home_stats, away_stats]).groupby('team_api_id').sum().reset_index()
    
    # 4. Teamnamen toevoegen
    df_teams_names = pd.read_sql_query("SELECT team_api_id, team_long_name FROM Team", connection)
    ranglijst = ranglijst.merge(df_teams_names, on='team_api_id')
    
    return ranglijst.sort_values(by='points', ascending=False).reset_index(drop=True)


df_2015_2016 = genereer_competitie_ranglijst(conn, '2015/2016')

# Resultaat bekijken
display(df_2015_2016[['team_long_name', 'points']].head(10))

,team_long_name,points
0,Paris Saint-Germain,96
1,FC Barcelona,91
2,Juventus,91
3,Real Madrid CF,90
4,Atlético Madrid,88
5,FC Bayern Munich,88
6,SL Benfica,88
7,Celtic,86
8,Sporting CP,86
9,PSV,84
